### PR TITLE
Consequencer jan 24

### PR DESCRIPTION
**Pull request by gamecat69**
reduced number of calls to update screen to improve performance for incoming clocks < 7ms
added grids patterns
added methods to reduce hysteresis on ain - this reduces the number of ain reads too
added methods to reduce hysteresis on k1 and k2 - this reduces the number of knob reads too
added screen saver to improve performance
changed the way the mode is displayed - replaced M1,M2,M3 with Mr,Mp,Mc for easier reading
cleaned up some code comments
added constants for easier code reading
updated documentation
applied black linting

### CODE REVIEW

File: software/contrib/consequencer.md [hunk 1]
```diff
@@ -86,20 +97,38 @@ Example usage for these CV outputs, (this will depend on the inputs your drum mo
...
 2. Send a control voltage into the analogue input
 
 A fixed voltage will select a single pattern and varying voltage (e.g. an envelope or LFO) will smoothly cycle through patterns.
 
+# Grids Patterns
+
+The 25 different patterns from the Mutable Instruments Grids module were exported from the Grids code and added to the Consequencer.
+
+Access the Grids patterns by holding button 2 for 3-5 seconds and releasing.
+Grids mode is enabled a period '.' before the pattern number on the bottom right of the screen.
```
Comment (roryjamesallen) [inline software/contrib/consequencer.md:119]: Spelling/Rewording - I think this is supposed to be "When Grids mode is enabled, a period '.' is visible before the pattern number on the bottom right of the screen"

File: software/contrib/consequencer.md [hunk 2]
```diff
@@ -86,20 +97,38 @@ Example usage for these CV outputs, (this will depend on the inputs your drum mo
...
+
+Two different versions of each Grids pattern have been created. The first version (even numbers, including 0) are the equivilant of having all three Grids density knobs at 50%. The second version (odd numbers) are the equivilant of having all three Grids density knobs at 70%. An example is below:
```
Comment (roryjamesallen) [inline software/contrib/consequencer.md:121]: Spelling - "equivalent"

File: software/contrib/consequencer.py [hunk 3]
```diff
@@ -5,100 +5,90 @@
...
 
+# Change detection thresholds to trigger a value change
+KNOB_CHANGE_TOLERANCE = 0.999
+AIN_CHANGE_TOLERANCE = 1
+
+# AIN Input Mode display chars
+MODE_DISPLAY_CHARS = ["r", "p", "c"]
+
+# Wake the screen upon detecting input at ain?
+WAKE_SCREEN_ON_AIN_INPUT = False
```
Comment (roryjamesallen) [inline software/contrib/consequencer.py:52]: This seems to be contrary to the README which states "Example interactions are : ain input"

File: software/contrib/consequencer.py [hunk 4]
```diff
@@ -211,88 +221,200 @@ def clockTriggerEnd():
...
+            ):
                 self.step = 0
                 self.clock_step = 0
 
+            # Has the module been left along for a while, turn off the screen
+            if (
+                not self.screenOff
+                and ticks_diff(ticks_ms(), self.lastInteractionTimeMs) > self.screenOffTimeoutMs
+            ):
+                self.drawBlankScreen()
```
Comment (roryjamesallen) [inline software/contrib/consequencer.py:401]: Possible place to use @chrisib's screensaver class, not a requirement for this PR but in future might reduce the amount of code needed to implement a new screensaver

**Review by roryjamesallen — COMMENTED** [review]
Still yet to test the actual script so my comments are just limited to glancing at the README and the code but looks good so far!

File: software/contrib/consequencer.md (hunk 2 above)
Comment (gamecat69) [reply software/contrib/consequencer.md:121]: Good spot! Fixed.

File: software/contrib/consequencer.py (hunk 3 above)
Comment (gamecat69) [reply software/contrib/consequencer.py:52]: Good spot. I removed ain input as a valid interaction during further testing as it seemed to be more of an interference than a nice feature to have. Updated the doc.

File: software/contrib/consequencer.py (hunk 4 above)
Comment (gamecat69) [reply software/contrib/consequencer.py:401]: I did indeed get inspiration for the screen saver from @chrisib . However, I chose to implement a simpler version to meet my more minimal requirements.

Comment (roryjamesallen) [reply software/contrib/consequencer.py:401]: Sweet, sounds good to me!

File: software/contrib/consequencer.py (hunk 3 above)
Comment (roryjamesallen) [reply software/contrib/consequencer.py:52]: Nice

**Comment by roryjamesallen** [comment]
If this has been tested with the most recent firmware i.e. including the recent PR merges, then I'm happy to merge

**Comment by gamecat69** [comment]
In response to the testing comment. This has been tested with the latest firmware. However, I am unsure of the easiest way to find the recent PR merges you refer to. Any pointers?

**Comment by roryjamesallen** [comment]
europi.py

> In response to the testing comment. This has been tested with the latest firmware. However, I am unsure of the easiest way to find the recent PR merges you refer to. Any pointers?

europi.py was updated in #328 and #326, I think that's the only one that's been changed that might affect this PR, and even then the changes might not affect it, but just good to check with the most recent files in the repo as opposed to the most recent release given how long ago the release was

**Comment by gamecat69** [comment]
Thanks for the pointer. I have just tested with the latest EuroPi.py and
all functions seem to work well.

**Review by roryjamesallen — APPROVED** [review]

**Pull request merged**